src/tvheadend/Settings.cpp

```diff
@@ -0,0 +1,187 @@
+const int         Settings::DEFAULT_HTSP_PORT           = 9982;
+const std::string Settings::DEFAULT_USERNAME            = "";
+const std::string Settings::DEFAULT_PASSWORD            = "";
+const int         Settings::DEFAULT_CONNECT_TIMEOUT     = 10000; // millisecs
+const int         Settings::DEFAULT_RESPONSE_TIMEOUT    = 5000;  // millisecs
+const bool        Settings::DEFAULT_TRACE_DEBUG         = false;
+const bool        Settings::DEFAULT_ASYNC_EPG           = false;
+const int         Settings::DEFAULT_TOTAL_TUNERS        = 1;  // total tuners > 1 => predictive tuning active
+const int         Settings::DEFAULT_PRETUNER_CLOSEDELAY = 10; // secs
+const int         Settings::DEFAULT_AUTOREC_MAXDIFF     = 15; // mins. Maximum difference between real and approximate start time for auto recordings
+const int         Settings::DEFAULT_APPROX_TIME         = 0;  // mins
+
+std::string Settings::ReadStringSetting(const std::string &key, const std::string &def)
+{
+  char value[1024];
+  if (XBMC->GetSetting(key.c_str(), value))
+    return value;
+
+  return def;
+}
+
+int Settings::ReadIntSetting(const std::string &key, int def)
+{
+  int value;
+  if (XBMC->GetSetting(key.c_str(), &value))
+    return value;
+
+  return def;
+}
+
+bool Settings::ReadBoolSetting(const std::string &key, bool def)
+{
+  bool value;
+  if (XBMC->GetSetting(key.c_str(), &value))
+    return value;
+
+  return def;
+}
+
+void Settings::ReadSettings()
+{
+  /* Connection */
+  SetHostname(ReadStringSetting("host", DEFAULT_HOST));
+  SetPortHTSP(ReadIntSetting("htsp_port", DEFAULT_HTSP_PORT));
+  SetPortHTTP(ReadIntSetting("http_port", DEFAULT_HTTP_PORT));
+  SetUsername(ReadStringSetting("user", DEFAULT_USERNAME));
+  SetPassword(ReadStringSetting("pass", DEFAULT_PASSWORD));
+
+  /* Note: Timeouts in settings UI are defined in seconds but we expect them to be in milliseconds. */
+  SetConnectTimeout(ReadIntSetting("connect_timeout", DEFAULT_CONNECT_TIMEOUT / 1000) * 1000);
+  SetResponseTimeout(ReadIntSetting("response_timeout", DEFAULT_RESPONSE_TIMEOUT / 1000) * 1000);
+
+  /* Debug */
+  SetTraceDebug(ReadBoolSetting("trace_debug", DEFAULT_TRACE_DEBUG));
+
+  /* Data Transfer */
+  SetTraceDebug(ReadBoolSetting("epg_async", DEFAULT_ASYNC_EPG));
+
+  /* Predictive Tuning */
+  bool bPretunerEnabled = ReadBoolSetting("pretuner_enabled", false);
+  SetTotalTuners(bPretunerEnabled ? ReadIntSetting("total_tuners", DEFAULT_TOTAL_TUNERS) : 1);
+  SetPreTunerCloseDelay(bPretunerEnabled ? ReadIntSetting("pretuner_closedelay", DEFAULT_PRETUNER_CLOSEDELAY) : 0);
+
+  /* Auto recordings */
+  SetAutorecApproxTime(ReadIntSetting("autorec_approxtime", DEFAULT_APPROX_TIME));
+  SetAutorecMaxDiff(ReadIntSetting("autorec_maxdiff", DEFAULT_AUTOREC_MAXDIFF));
+}
+
+ADDON_STATUS Settings::SetStringSetting(const std::string &oldValue, const void *newValue)
+{
+  if (oldValue == std::string(reinterpret_cast<const char *>(newValue)))
+  {
+    // value not changed.
```

Comment:
Minor thing, but if you'd add these comments in the method documentation in the header file then this method and the two below it could be shortened to 4 lines each instead of the current 10.